readers/influxdb/messages.go

```diff
@@ -1,150 +1,186 @@
 package influxdb
 
 import (
+	"context"
 	"encoding/json"
 	"fmt"
-	"strconv"
+	"strings"
 	"time"
 	"unicode"
 
 	"github.com/mainflux/mainflux/pkg/errors"
 	"github.com/mainflux/mainflux/readers"
 
-	influxdata "github.com/influxdata/influxdb/client/v2"
+	influxdb2 "github.com/influxdata/influxdb-client-go/v2"
 	jsont "github.com/mainflux/mainflux/pkg/transformers/json"
 	"github.com/mainflux/mainflux/pkg/transformers/senml"
 )
 
 const (
-	countCol = "count_protocol"
 	// Measurement for SenML messages
 	defMeasurement = "messages"
 )
 
 var _ readers.MessageRepository = (*influxRepository)(nil)
 
 var (
-	errResultSet  = errors.New("invalid result set")
 	errResultTime = errors.New("invalid result time")
 )
 
+type RepoConfig struct {
+	Bucket string
+	Org    string
+}
 type influxRepository struct {
-	database string
-	client   influxdata.Client
+	cfg    RepoConfig
+	client influxdb2.Client
 }
 
 // New returns new InfluxDB reader.
-func New(client influxdata.Client, database string) readers.MessageRepository {
+func New(client influxdb2.Client, repoCfg RepoConfig) readers.MessageRepository {
 	return &influxRepository{
-		database,
+		repoCfg,
 		client,
 	}
 }
 
 func (repo *influxRepository) ReadAll(chanID string, rpm readers.PageMetadata) (readers.MessagesPage, error) {
+
 	format := defMeasurement
 	if rpm.Format != "" {
 		format = rpm.Format
 	}
 
-	condition := fmtCondition(chanID, rpm)
-
-	cmd := fmt.Sprintf(`SELECT * FROM %s WHERE %s ORDER BY time DESC LIMIT %d OFFSET %d`, format, condition, rpm.Limit, rpm.Offset)
-	q := influxdata.Query{
-		Command:  cmd,
-		Database: repo.database,
-	}
+	queryAPI := repo.client.QueryAPI(repo.cfg.Org)
+	var sb strings.Builder
 
-	resp, err := repo.client.Query(q)
+	condition, timeRange := fmtCondition(chanID, rpm)
+	sb.WriteString(`import "influxdata/influxdb/v1"`)
+	sb.WriteString(fmt.Sprintf(`from(bucket: "%s")`, repo.cfg.Bucket))
+	// FluxQL syntax requires timeRange filter in this position, do not change.
+	sb.WriteString(timeRange)
+	// This is required to get messsage structure. Otherwise query returns fields in seperate rows.
+	sb.WriteString(`|> v1.fieldsAsCols()`)
+	sb.WriteString(`|> group()`)
+	sb.WriteString(fmt.Sprintf(`|> filter(fn: (r) => r._measurement == "%s")`, format))
+	sb.WriteString(condition)
+	sb.WriteString(`|> sort(columns: ["_time"], desc: true)`)
+	sb.WriteString(fmt.Sprintf(`|> limit(n:%d,offset:%d)`, rpm.Limit, rpm.Offset))
+	sb.WriteString(`|> yield(name: "sort")`)
```

Comment:
Try to simplify. Something like a single `fmt.Sprintf` should work.

Reply:
I agree, but I love the string builder here. It makes it modular if we want to modify the query later. I also think it makes it easy to understand. FluxQL language is very suitable for string builder

What do you think? If you say I should change I can.

Reply:
```golang
query := fmt.Sprintf(`
	import "influxdata/influxdb/v1" from(bucket: "%s")
	%s
	|> v1.fieldsAsCols()
	|> group()
	|> filter(fn: (r) => r._measurement == "%s")
	%s
	|> sort(columns: ["_time"], desc: true)
	|> limit(n:%d,offset:%d)
	|> yield(name: "sort")`,
		repo.cfg.Bucket,
		timeRange,
		format,
		condition,
		rpm.Limit, rpm.Offset,
	)

```

Hi again, simplified like this.